File: net/crates/net/sdk-ts/src/capability-schema.ts
/**
 * Capability axis schema + `validateCapabilities` — Phase 9a of
 * `CAPABILITY_SYSTEM_SDK_PLAN.md`.
 *
 * Mirrors the substrate's `AXIS_SCHEMA` const + the canonical
 * `validate_capabilities` validator. The wire shape of the
 * `ValidationReport` (lowercase `kind` discriminator, axis as
 * lowercase string, value type as lowercase string) is pinned by the
 * cross-binding fixture `tests/cross_lang_capability/capability_validation.json`.
 *
 * Source-of-truth for the schema is
 * `net/crates/net/docs/CAPABILITIES_SCHEMA.md`. The substrate's Rust
 * mirror in `behavior/schema.rs` and this TS mirror are
 * hand-maintained against the doc; CI guards each binding's
 * regenerated schema once the codegen tool from Phase 9a's plan
 * lands.
 *
 * @packageDocumentation
 */

import type {
  CapabilitySetWire,
  TaxonomyAxis,
} from './capability-enhancements';
import { tagFromString } from './capability-enhancements';

// ============================================================================
// Schema vocabulary
// ============================================================================

export type ValueType =
  | 'presence'
  | 'number'
  | 'string'
  | 'enumeration'
  | 'bool'
  | 'csv';

export interface KeyEntry {
  /** Full key under its axis, e.g. `cpu_cores`, `gpu.vendor`. */
  key: string;
  valueType: ValueType;
}

export type KeyShapeKind =
  | { kind: 'indexedCollection' }
  | { kind: 'keyedMap'; valueType: ValueType };

export interface KeyShape {
  /** Axis-relative prefix, e.g. `model.` (trailing `.` mandatory). */
  prefix: string;
  shape: KeyShapeKind;
  /** Per-element sub-keys for indexed collections; empty for keyed maps. */
  subKeys: KeyEntry[];
}

export interface AxisEntry {
  keys: KeyEntry[];
  shapes: KeyShape[];
}

export interface AxisSchema {
  hardware: AxisEntry;
  software: AxisEntry;
  devices: AxisEntry;
  dataforts: AxisEntry;
  metadataReserved: string[];
  metadataReservedPrefixes: string[];
}

// ============================================================================
// AXIS_SCHEMA — mirrors `behavior::schema::AXIS_SCHEMA`.
// ============================================================================

const HARDWARE_KEYS: KeyEntry[] = [
  { key: 'cpu_cores', valueType: 'number' },
  { key: 'cpu_threads', valueType: 'number' },
  { key: 'memory_mb', valueType: 'number' },
  { key: 'gpu', valueType: 'presence' },
  { key: 'gpu.vendor', valueType: 'enumeration' },
  { key: 'gpu.model', valueType: 'string' },
  { key: 'gpu.vram_mb', valueType: 'number' },
  { key: 'gpu.compute_units', valueType: 'number' },
  { key: 'gpu.tensor_cores', valueType: 'number' },
  { key: 'gpu.fp16_tflops_x10', valueType: 'number' },
  { key: 'storage_mb', valueType: 'number' },
  { key: 'network_mbps', valueType: 'number' },
  { key: 'limits.max_concurrent_requests', valueType: 'number' },
  { key: 'limits.max_tokens_per_request', valueType: 'number' },
  { key: 'limits.rate_limit_rpm', valueType: 'number' },
  { key: 'limits.max_batch_size', valueType: 'number' },
  { key: 'limits.max_input_bytes', valueType: 'number' },
  { key: 'limits.max_output_bytes', valueType: 'number' },
];

const SOFTWARE_KEYS: KeyEntry[] = [
  { key: 'os', valueType: 'string' },
  { key: 'os_version', valueType: 'string' },
  { key: 'cuda_version', valueType: 'string' },
];

const SOFTWARE_SHAPES: KeyShape[] = [
  {
    prefix: 'runtime.',
    shape: { kind: 'keyedMap', valueType: 'string' },
    subKeys: [],
  },
  {
    prefix: 'framework.',
    shape: { kind: 'keyedMap', valueType: 'string' },
    subKeys: [],
  },
  {
    prefix: 'driver.',
    shape: { kind: 'keyedMap', valueType: 'string' },
    subKeys: [],
  },
  {
    prefix: 'model.',
    shape: { kind: 'indexedCollection' },
    subKeys: [
      { key: 'id', valueType: 'string' },
      { key: 'family', valueType: 'string' },
      { key: 'parameters_b_x10', valueType: 'number' },
      { key: 'context_length', valueType: 'number' },
      { key: 'quantization', valueType: 'string' },
      { key: 'modalities', valueType: 'csv' },
      { key: 'tokens_per_sec', valueType: 'number' },
      { key: 'loaded', valueType: 'bool' },
    ],
  },
  {
    prefix: 'tool.',
    shape: { kind: 'indexedCollection' },
    subKeys: [
      { key: 'tool_id', valueType: 'string' },
      { key: 'name', valueType: 'string' },
      { key: 'version', valueType: 'string' },
      { key: 'requires', valueType: 'csv' },
      { key: 'estimated_time_ms', valueType: 'number' },
      { key: 'stateless', valueType: 'bool' },
    ],
  },
];

/** Reserved metadata keys (substrate-defined). */
export const METADATA_RESERVED_KEYS: readonly string[] = [
  'intent',
  'colocate-with',
  'colocate-with-strict',
  'priority',
  'owner',
];

/** Reserved metadata-key prefixes (`tool::<id>::input_schema` etc.). */
export const METADATA_RESERVED_PREFIXES: readonly string[] = ['tool::'];

/** Default soft cap for `metadata` total size. */
export const METADATA_SOFT_CAP_BYTES = 4 * 1024;

/** The canonical axis schema. Mirrors `behavior::schema::AXIS_SCHEMA`. */
export const AXIS_SCHEMA: AxisSchema = {
  hardware: { keys: HARDWARE_KEYS, shapes: [] },
  software: { keys: SOFTWARE_KEYS, shapes: SOFTWARE_SHAPES },
  devices: { keys: [], shapes: [] },
  dataforts: { keys: [], shapes: [] },
  metadataReserved: [...METADATA_RESERVED_KEYS],
  metadataReservedPrefixes: [...METADATA_RESERVED_PREFIXES],
};

// ============================================================================
// ValidationReport wire types
// ============================================================================

export type SchemaError =
  | {
      kind: 'unknown_axis';
      axis_prefix: string;
      tag: string;
    }
  | {
      kind: 'type_mismatch';
      axis: TaxonomyAxis;
      key: string;
      expected: ValueType;
      actual: string;
    }
  | {
      kind: 'index_malformed';
      axis: TaxonomyAxis;
      prefix: string;
      index: string;
      tag: string;
    };

export type ValidationWarning =
  | {
      kind: 'unknown_key';
      axis: TaxonomyAxis;
      key: string;
    }
  | {
      kind: 'metadata_oversize';
      soft_cap_bytes: number;
      actual_bytes: number;
    }
  | {
      kind: 'legacy_tag';
      tag: string;
    }
  // P2-H: mirror the substrate's CR-14 reserved-metadata warnings
  // (`schema.rs::ValidationWarning::MetadataReservedKey` /
  // `MetadataReservedPrefix`). Cross-binding parity restored.
  | {
      kind: 'metadata_reserved_key';
      key: string;
    }
  | {
      kind: 'metadata_reserved_prefix';
      key: string;
      prefix: string;
    };

export interface ValidationReport {
  errors: SchemaError[];
  warnings: ValidationWarning[];
}

// ============================================================================
// Validator
// ============================================================================

function axisEntry(schema: AxisSchema, axis: TaxonomyAxis): AxisEntry {
  switch (axis) {
    case 'hardware':
      return schema.hardware;
    case 'software':
      return schema.software;
    case 'devices':
      return schema.devices;
    case 'dataforts':
      return schema.dataforts;
  }
}

function checkValue(
  entry: KeyEntry,
  observedType: ValueType,
  observedValue: string | undefined,
  axis: TaxonomyAxis,
  errors: SchemaError[],
): void {
  if (entry.valueType === 'presence') {
    if (observedType !== 'presence') {
      errors.push({
        kind: 'type_mismatch',
        axis,
        key: entry.key,
        expected: 'presence',
        actual: observedValue ?? '',
      });
    }
    return;
  }
  if (observedValue === undefined) {
    errors.push({
      kind: 'type_mismatch',
      axis,
      key: entry.key,
      expected: entry.valueType,
      actual: '<no value>',
    });
    return;
  }
  let parses = false;
  switch (entry.valueType) {
    case 'number':
      // Substrate `Number` is unsigned (u64-only) — see CR-15 in the
      // capability-system-2 review and `schema.rs::ValueType::Number`.
      // Negative values surface as `TypeMismatch` errors on the
      // substrate side; mirror that here so client-side validation
      // doesn't flag a CapabilitySet as valid that the substrate
      // would later reject.
      parses = /^\d+$/.test(observedValue);
      break;
    case 'string':
    case 'enumeration':
    case 'csv':
      parses = observedValue.length > 0;
      break;
    case 'bool':
      parses = observedValue === 'true' || observedValue === 'false';
      break;
  }
  if (!parses) {
    errors.push({
      kind: 'type_mismatch',
      axis,
      key: entry.key,
      expected: entry.valueType,
      actual: observedValue,
    });
  }
}

function validateAxisKey(
  axis: TaxonomyAxis,
  key: string,
  observedType: ValueType,
  observedValue: string | undefined,
  schema: AxisSchema,
  errors: SchemaError[],
  warnings: ValidationWarning[],
  tagWire: string,
): void {
  const entry = axisEntry(schema, axis);
  const fixed = entry.keys.find((e) => e.key === key);
  if (fixed) {
    checkValue(fixed, observedType, observedValue, axis, errors);
    return;
  }
  for (const shape of entry.shapes) {
    if (!key.startsWith(shape.prefix)) continue;
    const rest = key.slice(shape.prefix.length);
    if (shape.shape.kind === 'indexedCollection') {
      const dot = rest.indexOf('.');
      if (dot < 0) continue;
      const idx = rest.slice(0, dot);
      const sub = rest.slice(dot + 1);
      // Q10: substrate parses the index as `u32`; strings of digits
      // longer than `u32::MAX` (e.g. `"4294967296"`, 2^32) are
      // accepted by the TS regex but rejected by the substrate as
      // `IndexMalformed`. Mirror the u32 range so client-side
      // validation doesn't silently pass payloads the substrate
      // would later reject.
      const idxNum = Number(idx);
      if (
        !/^\d+$/.test(idx) ||
        !Number.isInteger(idxNum) ||
        idxNum > 0xffff_ffff
      ) {
        errors.push({
          kind: 'index_malformed',
          axis,
          prefix: shape.prefix,
          index: idx,
          tag: tagWire,
        });
        return;
      }
      const subEntry = shape.subKeys.find((e) => e.key === sub);
      if (subEntry) {
        checkValue(subEntry, observedType, observedValue, axis, errors);
        return;
      }
      warnings.push({ kind: 'unknown_key', axis, key });
      return;
    }
    // KeyedMap: the rest IS the user-defined name.
    if (rest.length > 0) {
      const synth: KeyEntry = {
        key: shape.prefix,
        valueType: shape.shape.valueType,
      };
      checkValue(synth, observedType, observedValue, axis, errors);
      return;
    }
  }
  warnings.push({ kind: 'unknown_key', axis, key });
}

/**
 * Validate a wire-format capability set against a schema. Defaults to
 * the canonical {@link AXIS_SCHEMA}; pass a custom schema for
 * application-specific extensions.
 *
 * Mirrors the substrate's `validate_capabilities`. Output shape pinned
 * by `tests/cross_lang_capability/capability_validation.json`.
 */
export function validateCapabilities(
  caps: CapabilitySetWire,
  schema: AxisSchema = AXIS_SCHEMA,
): ValidationReport {
  const errors: SchemaError[] = [];
  const warnings: ValidationWarning[] = [];

  for (const wire of caps.tags) {
    const tag = tagFromString(wire);
    switch (tag.kind) {
      case 'axisPresent':
        validateAxisKey(
          tag.axis,
          tag.key,
          'presence',
          undefined,
          schema,
          errors,
          warnings,
          wire,
        );
        break;
      case 'axisValue':
        validateAxisKey(
          tag.axis,
          tag.key,
          'string',
          tag.value,
          schema,
          errors,
          warnings,
          wire,
        );
        break;
      case 'reserved':
        // Reserved-prefix tags pass through unchecked.
        break;
      case 'legacy':
        warnings.push({ kind: 'legacy_tag', tag: tag.raw });
        break;
    }
  }

  // P2-H: metadata-key reservation check. The schema declares
  // `metadataReserved` (exact-match) and `metadataReservedPrefixes`
  // (prefix-match) but pre-fix the validator never consulted them
  // — a user's `with_metadata("intent", …)` smuggling onto a
  // scheduler-reserved key emitted no warning. Mirrors the
  // substrate's CR-14 fix.
  for (const key of Object.keys(caps.metadata)) {
    if (schema.metadataReserved.includes(key)) {
      warnings.push({ kind: 'metadata_reserved_key', key });
      continue;
    }
    const prefix = schema.metadataReservedPrefixes.find((p) =>
      key.startsWith(p),
    );
    if (prefix !== undefined) {
      warnings.push({ kind: 'metadata_reserved_prefix', key, prefix });
    }
  }

  // Metadata soft-cap check.
  let metadataBytes = 0;
  for (const [k, v] of Object.entries(caps.metadata)) {
    metadataBytes += k.length + v.length;
  }
  if (metadataBytes > METADATA_SOFT_CAP_BYTES) {
    warnings.push({
      kind: 'metadata_oversize',
      soft_cap_bytes: METADATA_SOFT_CAP_BYTES,
      actual_bytes: metadataBytes,
    });
  }

  return { errors, warnings };
}

/** True iff there are zero errors and zero warnings. */
export function isReportClean(r: ValidationReport): boolean {
  return r.errors.length === 0 && r.warnings.length === 0;
}

/** True iff there are zero errors. Warnings are allowed. */
export function isReportValid(r: ValidationReport): boolean {
  return r.errors.length === 0;
}

File: net/crates/net/bindings/go/net/capability_schema.go
// Capability axis schema + ValidateCapabilities — Phase 9a of
// `CAPABILITY_SYSTEM_SDK_PLAN.md`.
//
// Mirrors the substrate's `AXIS_SCHEMA` const + the canonical
// `validate_capabilities` validator. The wire shape of the
// `ValidationReport` (lowercase `kind` discriminator, axis as
// lowercase string, value type as lowercase string) is pinned by the
// cross-binding fixture
// `tests/cross_lang_capability/capability_validation.json`.
//
// Source-of-truth for the schema is
// `net/crates/net/docs/CAPABILITIES_SCHEMA.md`. This Go reference
// implementation is hand-maintained against the doc; the codegen
// guard from Phase 9a will land alongside the production go.mod.

package net

import (
	"strconv"
	"strings"
)

// ============================================================================
// Schema vocabulary
// ============================================================================

// ValueType discriminates the value shape of an axis key. Matches the
// canonical wire string ("presence", "number", "string", "enumeration",
// "bool", "csv").
type ValueType string

const (
	ValueTypePresence    ValueType = "presence"
	ValueTypeNumber      ValueType = "number"
	ValueTypeString      ValueType = "string"
	ValueTypeEnumeration ValueType = "enumeration"
	ValueTypeBool        ValueType = "bool"
	ValueTypeCsv         ValueType = "csv"
)

// SchemaKeyEntry describes a fixed key under an axis.
type SchemaKeyEntry struct {
	Key       string
	ValueType ValueType
}

// SchemaShapeKind discriminates KeyShape.
type SchemaShapeKind uint8

const (
	SchemaShapeIndexedCollection SchemaShapeKind = iota
	SchemaShapeKeyedMap
)

// SchemaKeyShape describes an indexed / keyed sub-namespace under an
// axis. For indexed collections, SubKeys enumerates the per-element
// sub-keys; for keyed maps, KeyedValueType is the value type of map
// entries.
type SchemaKeyShape struct {
	Prefix         string
	Kind           SchemaShapeKind
	SubKeys        []SchemaKeyEntry
	KeyedValueType ValueType
}

// AxisEntry holds the fixed keys + shape patterns for one axis.
type SchemaAxisEntry struct {
	Keys   []SchemaKeyEntry
	Shapes []SchemaKeyShape
}

// AxisSchema is the top-level schema bundle.
type AxisSchema struct {
	Hardware                 SchemaAxisEntry
	Software                 SchemaAxisEntry
	Devices                  SchemaAxisEntry
	Dataforts                SchemaAxisEntry
	MetadataReserved         []string
	MetadataReservedPrefixes []string
}

// ============================================================================
// AxisSchemaCanonical — mirrors `behavior::schema::AXIS_SCHEMA`.
// ============================================================================

var hardwareKeys = []SchemaKeyEntry{
	{"cpu_cores", ValueTypeNumber},
	{"cpu_threads", ValueTypeNumber},
	{"memory_gb", ValueTypeNumber},
	{"gpu", ValueTypePresence},
	{"gpu.vendor", ValueTypeEnumeration},
	{"gpu.model", ValueTypeString},
	{"gpu.vram_gb", ValueTypeNumber},
	{"gpu.compute_units", ValueTypeNumber},
	{"gpu.tensor_cores", ValueTypeNumber},
	{"gpu.fp16_tflops_x10", ValueTypeNumber},
	{"storage_gb", ValueTypeNumber},
	{"network_gbps", ValueTypeNumber},
	{"limits.max_concurrent_requests", ValueTypeNumber},
	{"limits.max_tokens_per_request", ValueTypeNumber},
	{"limits.rate_limit_rpm", ValueTypeNumber},
	{"limits.max_batch_size", ValueTypeNumber},
	{"limits.max_input_bytes", ValueTypeNumber},
	{"limits.max_output_bytes", ValueTypeNumber},
}

var softwareKeys = []SchemaKeyEntry{
	{"os", ValueTypeString},
	{"os_version", ValueTypeString},
	{"cuda_version", ValueTypeString},
}

var softwareShapes = []SchemaKeyShape{
	{Prefix: "runtime.", Kind: SchemaShapeKeyedMap, KeyedValueType: ValueTypeString},
	{Prefix: "framework.", Kind: SchemaShapeKeyedMap, KeyedValueType: ValueTypeString},
	{Prefix: "driver.", Kind: SchemaShapeKeyedMap, KeyedValueType: ValueTypeString},
	{
		Prefix: "model.",
		Kind:   SchemaShapeIndexedCollection,
		SubKeys: []SchemaKeyEntry{
			{"id", ValueTypeString},
			{"family", ValueTypeString},
			{"parameters_b_x10", ValueTypeNumber},
			{"context_length", ValueTypeNumber},
			{"quantization", ValueTypeString},
			{"modalities", ValueTypeCsv},
			{"tokens_per_sec", ValueTypeNumber},
			{"loaded", ValueTypeBool},
		},
	},
	{
		Prefix: "tool.",
		Kind:   SchemaShapeIndexedCollection,
		SubKeys: []SchemaKeyEntry{
			{"tool_id", ValueTypeString},
			{"name", ValueTypeString},
			{"version", ValueTypeString},
			{"requires", ValueTypeCsv},
			{"estimated_time_ms", ValueTypeNumber},
			{"stateless", ValueTypeBool},
		},
	},
}

// MetadataReservedKeys lists substrate-defined reserved metadata
// keys.
var MetadataReservedKeys = []string{
	"intent",
	"colocate-with",
	"colocate-with-strict",
	"priority",
	"owner",
}

// MetadataReservedPrefixes lists reserved metadata-key prefixes.
var MetadataReservedPrefixes = []string{"tool::"}

// MetadataSoftCapBytes is the default soft cap for `metadata` total
// size.
const MetadataSoftCapBytes = 4 * 1024

// AxisSchemaCanonical mirrors `behavior::schema::AXIS_SCHEMA`.
var AxisSchemaCanonical = AxisSchema{
	Hardware:                 SchemaAxisEntry{Keys: hardwareKeys, Shapes: nil},
	Software:                 SchemaAxisEntry{Keys: softwareKeys, Shapes: softwareShapes},
	Devices:                  SchemaAxisEntry{},
	Dataforts:                SchemaAxisEntry{},
	MetadataReserved:         MetadataReservedKeys,
	MetadataReservedPrefixes: MetadataReservedPrefixes,
}

// ============================================================================
// ValidationReport wire types
// ============================================================================

// SchemaError is the wire-format schema-violation record. Only one
// of the optional groups is populated per kind; JSON-omitempty on
// the unused fields produces the canonical wire form.
type SchemaError struct {
	Kind       string    `json:"kind"`
	AxisPrefix string    `json:"axis_prefix,omitempty"`
	Tag        string    `json:"tag,omitempty"`
	Axis       string    `json:"axis,omitempty"`
	Key        string    `json:"key,omitempty"`
	Expected   ValueType `json:"expected,omitempty"`
	Actual     string    `json:"actual,omitempty"`
	Prefix     string    `json:"prefix,omitempty"`
	Index      string    `json:"index,omitempty"`
}

// ValidationWarning is the wire-format forward-compat / hygiene
// record.
type ValidationWarning struct {
	Kind         string `json:"kind"`
	Axis         string `json:"axis,omitempty"`
	Key          string `json:"key,omitempty"`
	SoftCapBytes int    `json:"soft_cap_bytes,omitempty"`
	ActualBytes  int    `json:"actual_bytes,omitempty"`
	Tag          string `json:"tag,omitempty"`
	// Q11: mirror substrate CR-14 reserved-metadata warnings.
	// Wire shape matches `src/ffi/schema.rs::validation_warning_to_wire`
	// — `metadata_reserved_prefix` carries both `key` (full key
	// the user wrote) AND `prefix` (the reserved prefix that
	// matched). The `key` field is reused; `Prefix` is the new
	// JSON-tagged field below.
	Prefix string `json:"prefix,omitempty"`
}

// ValidationReport is the validator's output.
type ValidationReport struct {
	Errors   []SchemaError       `json:"errors"`
	Warnings []ValidationWarning `json:"warnings"`
}

// IsClean returns true iff there are zero errors and zero warnings.
func (r ValidationReport) IsClean() bool {
	return len(r.Errors) == 0 && len(r.Warnings) == 0
}

// IsValid returns true iff there are zero errors. Warnings are allowed.
func (r ValidationReport) IsValid() bool {
	return len(r.Errors) == 0
}

// ============================================================================
// Validator
// ============================================================================

func axisEntry(schema *AxisSchema, axis TaxonomyAxis) *SchemaAxisEntry {
	switch axis {
	case AxisHardware:
		return &schema.Hardware
	case AxisSoftware:
		return &schema.Software
	case AxisDevices:
		return &schema.Devices
	case AxisDataforts:
		return &schema.Dataforts
	}
	return nil
}

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

// N-5: substrate `ValueType::Number` is unsigned u64. Rust uses
// `value.parse::<u64>()` (schema.rs:704), which accepts `+1` and
// rejects negatives, non-ASCII digits, and values exceeding
// `u64::MAX`. Python locks the same accepted-set via the
// `_U64_LITERAL = re.compile(r"^\+?[0-9]+$")` regex plus an
// explicit ceiling check (R4 / capability_schema.py:332,385).
//
// The pre-fix Go shape used `isIntegerLiteral`, which admitted a
// leading `-` AND any number of digits — `software.model.0.context_length=-1`
// or `=18446744073709551616` would validate clean in Go and error
// in Rust/Python. Mirror Python: ASCII digits with optional leading
// `+`, then `strconv.ParseUint` for the range check.
func isU64Literal(s string) bool {
	if s == "" {
		return false
	}
	body := s
	if body[0] == '+' {
		body = body[1:]
	}
	if !isAllDigits(body) {
		return false
	}
	if _, err := strconv.ParseUint(body, 10, 64); err != nil {
		return false
	}
	return true
}

// N-5: substrate `schema.rs:616` parses the indexed-collection
// position as `idx.parse::<u32>()`. Pre-fix `isAllDigits` admitted
// `99999999999999` which overflows u32. Mirror Rust strictly.
func isU32Literal(s string) bool {
	if s == "" {
		return false
	}
	body := s
	if body[0] == '+' {
		body = body[1:]
	}
	if !isAllDigits(body) {
		return false
	}
	if _, err := strconv.ParseUint(body, 10, 32); err != nil {
		return false
	}
	return true
}

func checkValue(
	entry SchemaKeyEntry,
	observedType ValueType,
	observedValue *string,
	axis TaxonomyAxis,
	errors *[]SchemaError,
) {
	if entry.ValueType == ValueTypePresence {
		if observedType != ValueTypePresence {
			actual := ""
			if observedValue != nil {
				actual = *observedValue
			}
			*errors = append(*errors, SchemaError{
				Kind:     "type_mismatch",
				Axis:     string(axis),
				Key:      entry.Key,
				Expected: ValueTypePresence,
				Actual:   actual,
			})
		}
		return
	}
	if observedValue == nil {
		*errors = append(*errors, SchemaError{
			Kind:     "type_mismatch",
			Axis:     string(axis),
			Key:      entry.Key,
			Expected: entry.ValueType,
			Actual:   "<no value>",
		})
		return
	}
	v := *observedValue
	parses := false
	switch entry.ValueType {
	case ValueTypeNumber:
		parses = isU64Literal(v)
	case ValueTypeString, ValueTypeEnumeration, ValueTypeCsv:
		parses = v != ""
	case ValueTypeBool:
		parses = v == "true" || v == "false"
	}
	if !parses {
		*errors = append(*errors, SchemaError{
			Kind:     "type_mismatch",
			Axis:     string(axis),
			Key:      entry.Key,
			Expected: entry.ValueType,
			Actual:   v,
		})
	}
}

func validateAxisKey(
	axis TaxonomyAxis,
	key string,
	observedType ValueType,
	observedValue *string,
	schema *AxisSchema,
	errors *[]SchemaError,
	warnings *[]ValidationWarning,
	tagWire string,
) {
	entry := axisEntry(schema, axis)
	if entry == nil {
		return
	}
	for _, e := range entry.Keys {
		if e.Key == key {
			checkValue(e, observedType, observedValue, axis, errors)
			return
		}
	}
	for _, shape := range entry.Shapes {
		if !strings.HasPrefix(key, shape.Prefix) {
			continue
		}
		rest := key[len(shape.Prefix):]
		switch shape.Kind {
		case SchemaShapeIndexedCollection:
			dot := strings.IndexByte(rest, '.')
			if dot < 0 {
				continue
			}
			idx := rest[:dot]
			sub := rest[dot+1:]
			if !isU32Literal(idx) {
				*errors = append(*errors, SchemaError{
					Kind:   "index_malformed",
					Axis:   string(axis),
					Prefix: shape.Prefix,
					Index:  idx,
					Tag:    tagWire,
				})
				return
			}
			for _, sk := range shape.SubKeys {
				if sk.Key == sub {
					checkValue(sk, observedType, observedValue, axis, errors)
					return
				}
			}
			*warnings = append(*warnings, ValidationWarning{
				Kind: "unknown_key",
				Axis: string(axis),
				Key:  key,
			})
			return
		case SchemaShapeKeyedMap:
			if rest != "" {
				synth := SchemaKeyEntry{Key: shape.Prefix, ValueType: shape.KeyedValueType}
				checkValue(synth, observedType, observedValue, axis, errors)
				return
			}
		}
	}
	*warnings = append(*warnings, ValidationWarning{
		Kind: "unknown_key",
		Axis: string(axis),
		Key:  key,
	})
}

// ValidateCapabilities runs the canonical validator against the
// canonical AxisSchemaCanonical.
func ValidateCapabilities(caps CapabilitySetWire) ValidationReport {
	return ValidateCapabilitiesAgainst(caps, &AxisSchemaCanonical)
}

// ValidateCapabilitiesAgainst runs the validator against a custom schema.
func ValidateCapabilitiesAgainst(
	caps CapabilitySetWire,
	schema *AxisSchema,
) ValidationReport {
	// R3: nil schema → fall back to the canonical schema. The
	// validateAxisKey / metadata-reservation walks below all
	// dereference `schema`; passing nil would panic. Coercing
	// to the canonical schema matches what `ValidateCapabilities`
	// (the public entry point with no schema arg) does and gives
	// callers who pass nil the same shape as the no-arg variant.
	if schema == nil {
		schema = &AxisSchemaCanonical
	}
	errors := []SchemaError{}
	warnings := []ValidationWarning{}

	for _, wire := range caps.Tags {
		tag, err := TagFromString(wire)
		if err != nil {
			continue
		}
		switch tag.Kind {
		case TagKindAxisPresent:
			validateAxisKey(
				tag.Axis, tag.Key, ValueTypePresence, nil,
				schema, &errors, &warnings, wire,
			)
		case TagKindAxisValue:
			v := tag.Value
			validateAxisKey(
				tag.Axis, tag.Key, ValueTypeString, &v,
				schema, &errors, &warnings, wire,
			)
		case TagKindReserved:
			// Reserved-prefix tags pass through unchecked.
		case TagKindLegacy:
			warnings = append(warnings, ValidationWarning{
				Kind: "legacy_tag",
				Tag:  tag.Raw,
			})
		}
	}

	// Q11: metadata-key reservation check. Mirror substrate CR-14
	// (and TS P2-H, Py P2-L). The schema declares
	// `MetadataReserved` (exact-match) and `MetadataReservedPrefixes`
	// (prefix-match) but pre-fix the Go validator never consulted
	// either list — a user's `WithMetadata("intent", …)` smuggling
	// onto a scheduler-reserved key emitted no warning client-side
	// even though Rust / TS / Python now flag it.
	for k := range caps.Metadata {
		matched := false
		for _, reserved := range schema.MetadataReserved {
			if k == reserved {
				warnings = append(warnings, ValidationWarning{
					Kind: "metadata_reserved_key",
					Key:  k,
				})
				matched = true
				break
			}
		}
		if matched {
			continue
		}
		for _, prefix := range schema.MetadataReservedPrefixes {
			if strings.HasPrefix(k, prefix) {
				warnings = append(warnings, ValidationWarning{
					Kind:   "metadata_reserved_prefix",
					Key:    k,
					Prefix: prefix,
				})
				break
			}
		}
	}

	metadataBytes := 0
	for k, v := range caps.Metadata {
		metadataBytes += len(k) + len(v)
	}
	if metadataBytes > MetadataSoftCapBytes {
		warnings = append(warnings, ValidationWarning{
			Kind:         "metadata_oversize",
			SoftCapBytes: MetadataSoftCapBytes,
			ActualBytes:  metadataBytes,
		})
	}

	return ValidationReport{Errors: errors, Warnings: warnings}
}
